### ISSUE TITLE
NTXentLoss with Miner

### ISSUE THREAD

**Issue by melgor** [KevinMusgrave/pytorch-metric-learning#196]
Hi, I have question regarding `NTXentLoss`. Exactly its connection to miner. In MoCo and [Instance Cross Entropy for Deep Metric Learning](https://arxiv.org/abs/1911.09976), there is no miner. I mean that for given pair, use all other negatives.
I can not get how does it work with ex. `DistanceWeightedMiner`. Do we care about negatives pairs here at all? I mean that in this mainer we take all possible positive pairs and for each one we choose negative pairs following the uniform distribution.
And then how it will impact this loss?
1. Take all negatives, ignoring somehow uniform distribution?
2. Will take only this single negative?

For me, it is hard to interpret the [code](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/src/pytorch_metric_learning/losses/ntxent_loss.py), without data-structures etc. It is not similar to the implementation which assume only 1 single positive pair in batch, so it also make it harder (but is awesome that it work with multiple instances from case class)

**Comment by melgor** [comment]
Look like I have found the answer: https://github.com/KevinMusgrave/pytorch-metric-learning/issues/6#issuecomment-587331886

So I will take all negatives which are associated with anchor. Then having such miner we are less element in denominator.
Look like a reasonable choice for this loss. And from my understanding, no loss make it work similar to SimClr, where all negatives are in denominator.

**Comment by KevinMusgrave** [comment]
> So I will take all negatives which are associated with anchor. Then having such miner we are less element in denominator.
> Look like a reasonable choice for this loss. And from my understanding, no loss make it work similar to SimClr, where all negatives are in denominator.

Yes, if you don't use a miner, then it will be like SimClr, but you can have more than 1 positive pair in the batch.

**Comment by melgor** [comment]
Thanks for answer!
